react-native-git-upgrade/cliEntry.js

```diff
@@ -0,0 +1,239 @@
+
+/**
+ * Promisify the callback-based shelljs function exec
+ * @param command
+ * @param opts
+ * @returns {Promise}
+ */
+function exec(context, command, forceOutput) {
+  return new Promise((resolve, reject) => {
+    shell.exec(command, {silent: (forceOutput ? !forceOutput : !context.cliArgs.verbose)}, (code, stdout, stderr) => {
+      code
+        ? reject(new Error(`Command '${command}' exited with code ${code}:
+stderr: ${stderr}
+stdout: ${stdout}`))
+        : resolve(stdout);
+    });
+  });
+}
+
+function readProjectDescriptors() {
+  try {
+    const rnPakPath = path.resolve(
+      process.cwd(),
+      'node_modules',
+      'react-native',
+      'package.json'
+    );
+
+    const pakPath = path.resolve(
+      process.cwd(),
+      'package.json'
+    );
+
+    const rnPak = JSON.parse(fs.readFileSync(rnPakPath, 'utf8'));
+    const pak = JSON.parse(fs.readFileSync(pakPath, 'utf8'));
+
+    return {rnPak, pak};
+  } catch (err) {
+    throw new Error(
+      'Unable to find project descriptors. Make sure that you have run `npm install` ' +
+      'and that you are inside a react-native project.'
+    )
+  }
+}
+
+
+function setupWorkingDir(context) {
+  return new Promise((resolve, reject) => {
+    rimraf(context.tmpDir, err => {
+      if (err) {
+        reject(err);
+      } else {
+        fs.mkdirSync(context.tmpDir);
+        resolve()
+      }
+    });
+  });
+}
+
+function configureGitEnv(context) {
+  process.env.GIT_DIR = path.resolve(context.tmpDir, '.gitrn');
+  process.env.GIT_WORK_TREE = '.';
+}
+
+function generateTemplates(context) {
+  try {
+    // Try requiring the index.js (entry-point of Yeoman generators)
+    const generatorEntryPoint = path.resolve(context.generatorDir, 'index.js');
+    fs.accessSync(generatorEntryPoint);
+    return runYeomanGenerators(context);
+  } catch(err) {
+    return runCopyAndReplace(context);
+  }
+}
+
+function runCopyAndReplace(context) {
+  const copyProjectTemplateAndReplacePath = path.resolve(context.generatorDir, 'copyProjectTemplateAndReplace');
+  // This module can be required twice in different RN version
+  delete require.cache[copyProjectTemplateAndReplacePath];
+  const copyProjectTemplateAndReplace = require(copyProjectTemplateAndReplacePath);
+  copyProjectTemplateAndReplace(
+    path.resolve(context.generatorDir, '..', 'templates', 'HelloWorld'),
+    process.cwd(),
+    context.appName,
+    {upgrade: true, force: true}
+  );
+}
+
+function runYeomanGenerators(context) {
+  if (!context.cliArgs.verbose) {
+    // Yeoman output needs monkey-patching (no silent option)
+    TerminalAdapter.prototype.log = () => {};
+    TerminalAdapter.prototype.log.force = () => {};
+  }
+
+  const env = yeoman.createEnv();
+  env.register(context.generatorDir, 'react:app');
+  const generatorArgs = ['react:app', context.appName].concat(context.cliArgs._);
+  return new Promise((resolve) => env.run(generatorArgs, {upgrade: true, force: true}, resolve));
```

Comment:
I wonder if it would be possible to simply call into the CLI using `require('node_modules/react-native/upgrade')` or similar and calling it. That way you might not need these few lines to use the Yeoman API directly?

Not a big deal.